tweet_recommendations/our_method.py

```diff
@@ -0,0 +1,286 @@
+            ranking = np.argsort(-sim_pop[i, :])
+            result.append(list(self._hashtags_df['hashtag'].iloc[ranking]))
+        return result
+
+    def normalise(self, array: np.ndarray):
+        if not np.isclose(array.max(), array.min(), rtol=1e-8):
+            return (array - array.min()) / (array.max() - array.min())
+        return array
+
+    def embedding_similarity(self, x: np.ndarray, y: np.ndarray):
+        """
+        Computes angular similarity based on cosine similarity.
+        https://en.wikipedia.org/wiki/Cosine_similarity#Angular_distance_and_similarity
+        Angular similarity is bound to [0, 1] and angular distance is a formal distance metric.
+        """
+
+        # clipping to avoid computational errors
+        similarity = np.clip(cosine_similarity(x, y), -1, 1)
+        ang_dist = np.arccos(similarity) / np.pi
+        ang_sim = 1 - ang_dist
+        return ang_sim
+
+    def _build_base_graph(self, tweets: pd.DataFrame):
+        """
+        Builds base tweets-hasthtags graph structure.
+        :param tweets: pd.DataFrame containing tweets & featured hashtags.
+        :return: nx.Graph of tweets & hashtags.
+        """
+
+        G = nx.Graph()
+
+        def add_row_to_graph(row, graph, embedding_name):
+            if len(row.hashtags) > 0:
+                graph.add_node(row.id)
+                graph.node[row.id]["node_type"] = "tweet"
+                graph.node[row.id]["retweets"] = row.retweet_count
+                graph.node[row.id][embedding_name] = row[embedding_name]
+
+                for hashtag in row.hashtags:
+                    graph.add_node(hashtag)
+                    graph.node[hashtag]["node_type"] = "hashtag"
+                    graph.add_edge(row.id, hashtag)
+
+        if self.verbose:
+            print("Building graph ...")
+            tqdm.tqdm.pandas()
+            tweets.progress_apply(lambda r:
+                                  add_row_to_graph(r, G, self.embedding_name),
+                                  axis=1)
+        else:
+            tweets.apply(lambda r: add_row_to_graph(r, G, self.embedding_name),
+                         axis=1)
+        return G
+
+    def _calculate_hashtag_embeddings(self, G: nx.Graph):
+
+        for node in tqdm.tqdm(G.nodes, total=len(G.nodes),
+                              disable=not self.verbose):
+            if G.nodes[node]["node_type"] == "hashtag":
+                tweets = list(G.neighbors(node))
+                embeddings = np.asarray([G.nodes[tweet][self.embedding_name]
+                                         for tweet in tweets])
+                G.nodes[node][self.embedding_name] = embeddings.mean(axis=0)
+
+        return G
+
+    def _calculate_pagerank(self, G: nx.Graph):
+        graph_pagerank = nx.pagerank(G)
+        nx.set_node_attributes(G, graph_pagerank, "pagerank")
+        return G
+
+    def _calculate_hashtag_mean_retweets(self, G: nx.Graph):

```

Comment:
Zależnie, jak duży jest graf, ale myślę, że dla optymalizacji można to włączyć do liczenia embeddingu `_calculate_hashtag_embeddings`. Zaoszczędzisz jedną iterację.